### PR TITLE
Bug fixes for MHA models

### PR DESCRIPTION
**Pull request by gau-nernst**
## Modifications

1. ~~**Cast bias to FP32 inside AMX GEMM** - AMX GEMM only accepts FP32 bias~~ **Convert bias to FP32 for sgl-kernel** since sgl-kernel only works with FP32. Do this in `prepack_weights_after_load()` so there is no runtime overhead
2. **Fix output shape of sgl-kernel decode attention** - output shape should have ndim=2 `(num_tokens, num_heads * head_dim)` instead of 3 `(num_tokens, num_heads, head_dim)`

I refer to Triton attention backend to see what's the expected input/output shape to attention
https://github.com/mingfeima/sglang/blob/aea281f654f6996029bf492d5c00d50bccae661f/python/sglang/srt/layers/attention/triton_backend.py#L428-L436

3. **Fix missing attribute when there is tied embedding** - When there is tied embedding, LM head (which is actually the embedding) doesn't have `use_intel_amx_backend` attribute -> causes error.

https://github.com/mingfeima/sglang/blob/aea281f654f6996029bf492d5c00d50bccae661f/python/sglang/srt/layers/logits_processor.py#L418-L419

Using `getattr()` might have some perf penalty. Hence, I add code to create `use_intel_amx_backend` on Embedding layer instead.

To test

```bash
# with tied embeddings
python -m sglang.bench_one_batch --batch-size 1 --input 1024 --output 1024 --model Qwen/Qwen2.5-0.5B-Instruct --device cpu

# w/o tied embeddings
python -m sglang.bench_one_batch --batch-size 1 --input 1024 --output 1024 --model Qwen/Qwen2.5-7B-Instruct --device cpu
```

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/attention/intel_amx_backend.py [hunk 1]
```diff
@@ -99,19 +99,18 @@ def forward_decode(
     ):
         attn_logits, _ = self.forward_metadata
 
-        if q.ndim == 2:
-            q = q.view(-1, layer.tp_q_head_num, layer.qk_head_dim)
+        q = q.reshape(-1, layer.tp_q_head_num * layer.qk_head_dim)
 
         if layer.qk_head_dim != layer.v_head_dim:
-            o = q.new_empty((q.shape[0], layer.tp_q_head_num, layer.v_head_dim))
+            o = q.new_empty((q.shape[0], layer.tp_q_head_num * layer.v_head_dim))
```
Comment (mingfeima) [inline python/sglang/srt/layers/attention/intel_amx_backend.py:105]: not sure if we can change this without blocking other models that we have. @yanbing-j  please comment.

File: sgl-kernel/csrc/cpu/gemm.cpp [hunk 2]
```diff
@@ -439,7 +439,7 @@ at::Tensor weight_packed_linear(at::Tensor& mat1, at::Tensor& mat2,
   const float* bias_data = nullptr;
   if (has_bias) {
     CHECK_EQ(bias.value().size(0), N);
-    bias_data = bias.value().data_ptr<float>();
+    bias_data = bias.value().to(at::kFloat).data_ptr<float>();
```
Comment (mingfeima) [inline sgl-kernel/csrc/cpu/gemm.cpp:442]: change `bias` from bfloat16 to float when loading the model would be a better idea as this won't hurt the performance.

**Review by mingfeima — CHANGES_REQUESTED** [review]
need comments from @yanbing-j

we haven't done model coverage test yet - feel free to raise issue :)

File: python/sglang/srt/layers/attention/intel_amx_backend.py (hunk 1 above)
Comment (yanbing-j) [reply python/sglang/srt/layers/attention/intel_amx_backend.py:105]: Meanwhile, we only test MHA part when we fix num_prompt. At that time, the above change has in cpu_opt_ww11. And we didn't run fail.

Comment (gau-nernst) [reply python/sglang/srt/layers/attention/intel_amx_backend.py:105]: This is interesting.. The problem I encountered for `Qwen/Qwen2.5-0.5B-Instruct` is that since the output of attention layer now is 3D, the next linear layer (output projection) will fail. Maybe it's specific to the model code, will investigate further.

Comment (gau-nernst) [reply python/sglang/srt/layers/attention/intel_amx_backend.py:105]: The reason 3D `o` was working for DeepSeek-V2/V3 because there is an explicit reshape after attention (MLA)

https://github.com/mingfeima/sglang/blob/d59d352330fa295671a30b6aaa644fccbe7477e0/python/sglang/srt/models/deepseek_v2.py#L430-L433

This is not the case for most other models (MHA/GQA)

https://github.com/mingfeima/sglang/blob/d59d352330fa295671a30b6aaa644fccbe7477e0/python/sglang/srt/models/llama.py#L181-L182

https://github.com/mingfeima/sglang/blob/d59d352330fa295671a30b6aaa644fccbe7477e0/python/sglang/srt/models/qwen2.py#L169-L170

If we look at `forward_decode()` of other attention backends, they also reshape `o` to 2D at the end

https://github.com/mingfeima/sglang/blob/d59d352330fa295671a30b6aaa644fccbe7477e0/python/sglang/srt/layers/attention/flashinfer_mla_backend.py#L396-L403

Hence, the conclusion is that
- Attention backend should return 2D `o`
- If we create 2D `o`, we will need `.view()` **before** calling attention kernel. If we create 3D `o`, we will need `.view()` **after** calling attention kernel -> performance-wise, it should be the same anyway

Do let me know how you want to proceed.

File: sgl-kernel/csrc/cpu/gemm.cpp (hunk 2 above)
Comment (gau-nernst) [reply sgl-kernel/csrc/cpu/gemm.cpp]: I move the code to `_process_weight_after_loading()`. Let me know if it addresses the problem.

**Comment by gau-nernst** [comment]
@yanbing-j Can we merge this PR first before #43? Because #43 has some changes from this PR.

**Review by yanbing-j — APPROVED** [review]
